cmd/gardenlet/app/app.go

```diff
@@ -557,6 +568,93 @@ func (g *garden) cleanupStaleShootStates(ctx context.Context, gardenClient clien
 	return flow.Parallel(taskFns...)(ctx)
 }
 
+func (g *garden) cleanupStaleAddonsMR(ctx context.Context, gardenClient, seedClient client.Client) error {
+	shootNamespaceList := &corev1.NamespaceList{}
+	if err := seedClient.List(ctx, shootNamespaceList, client.MatchingLabels{v1beta1constants.GardenRole: v1beta1constants.GardenRoleShoot}); err != nil {
+		return err
+	}
+
+	var taskFns []flow.TaskFn
+
+	for _, ns := range shootNamespaceList.Items {
+		namespace := ns
+
+		taskFns = append(taskFns, func(ctx context.Context) error {
+			var (
+				addonsMR = &resourcesv1alpha1.ManagedResource{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "addons",
+						Namespace: namespace.Name,
+					},
+				}
+				addonsSecret = &corev1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "managedresource-addons",
+						Namespace: namespace.Name,
+					},
+				}
+				resourceManagerDeployment = &appsv1.Deployment{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      v1beta1constants.DeploymentNameGardenerResourceManager,
+						Namespace: namespace.Name,
+					},
+				}
+			)
+
+			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(addonsMR), addonsMR); err != nil {
+				// If the MR is already gone, then nothing to do here
+				if apierrors.IsNotFound(err) {
+					return nil
+				}
+				return err
+			}
+
+			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(resourceManagerDeployment), resourceManagerDeployment); err != nil {
```

Comment:
Just to be on the safe side:
```suggestion
			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(resourceManagerDeployment), resourceManagerDeployment); err != nil {
				if apierrors.IsNotFound(err) {
					return nil
				}
```